rsconf/component/sirepo_sandbox.py

```diff
@@ -0,0 +1,124 @@
+                self.buildt.build_component(t)
+            return
+        self.buildt.require_component("docker", "network")
+        jc, _ = self.j2_ctx_init()
+        z = jc.sirepo_sandbox
+        z.setdefault("volumes", {})
+        # TODO(e-carlin): Better way then building the path manually
+        # /srv/<component-name> is output by systemd.docker_unit_prepare
+        # below but to call that we need z.host.ssh_d which needs z.host_d
+        z.host_d = z.host_root_d.join(self.name, "db")
+        z.secrets = self._gen_keys(jc, z)
+        for x in "guest", "host":
+            z[x] = self._gen_paths(z, z.get(x + "_d"))
+        z.run_u = jc.rsconf_db.run_u
+        # Only additional config for the server is the sshd config.
+        z.run_d = systemd.docker_unit_prepare(self, jc, watch_files=[z.host.ssh_d])
+        self._network(jc, z)
+
+    def internal_build_write(self):
+        from rsconf import systemd
+
+        if self.name == "sirepo_sandbox":
+            self.append_root_bash(": nothing to do")
+            return
+        jc = self.j2_ctx
+        z = jc.sirepo_sandbox
+        v = []
+        x = [
+            [z.host.ssh_d, z.guest.ssh_d],
+            # SECURITY: no modifications to run_d
+            [z.run_d, z.run_d, "ro"],
+        ]
+        for d in z.volumes:
+            h = z.host[d]
+            x.append([h, z.guest[d]])
+            v.append(h)
+        systemd.docker_unit_enable(
+            self,
+            jc,
+            image=jc.sirepo_sandbox.docker_image,
+            volumes=x,
+            cmd="/usr/sbin/sshd -D -f '{}'".format(z.guest.sshd_config),
+        )
+        self.install_access(mode="700", owner=z.run_u)
+        for d in [z.host_d, z.host.ssh_d, *v]:
+            self.install_directory(d)
+        self.install_access(mode="400")
+        self.install_resource(z.host.sshd_config, jc)
+        for k, v in z.secrets.items():
+            self.install_abspath(v, z.host[k])
+
+    def _network(self, jc, z):
+        n = self.buildt.get_component("network")
+        z.ip, _ = n.ip_and_net_for_host(jc.rsconf_db.host)
+        z.ssh_port = jc.sirepo_sandbox.users[self.name]
+
+    def _gen_keys(self, jc, z):
+        from rsconf import db
+
+        res = PKDict()
+        b = db.secret_path(
+            jc,
+            "sirepo_sandbox/" + self.name,
+            visibility="channel",
+        )
+        pkio.mkdir_parent(b)
+        res.host_key_f = b.join("host_key")
+        res.host_key_pub_f = res.host_key_f + ".pub"
+        res.identity_f = b.join("identity")
+        res.identity_pub_f = b.join("identity") + ".pub"
+        for f in res.host_key_f, res.identity_f:
+            if f.exists():
+                continue
+            subprocess.check_call(
+                [
+                    "ssh-keygen",
+                    "-q",
+                    "-t",
+                    "ed25519",
+                    "-N",
+                    z.user_passwords[self.name],
```

Comment:
You said:

> Already encrypted private keys will need to be distributed to end users outside the rsconf configured nodes. This is different than mpi_worker which uses non-encrypted keys and the distribution happens to external devices (laptops). The main point is the key is encrypted by sirepo_sandbox component with a passphrase that is configured and transferred to users along with the keys. The protocol can be manual as it is all happening inside the organization.

I wasn't 100% sure what you meant but is this it? We can manually create passwords and populate them in db/secret/000.yml, run `rsconf build`, then manually distribute the password/key to users?

Reply:
I think we can generate passwords and it will be simpler than configuring them. They do not need to be distributed, just dumped to json. Just like rsconf does, for example. These passwords will be changed by the user once the private key is on their machine.

Reply:
8d1bb90